src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs

```diff
@@ -163,12 +177,111 @@ protected override void ExecuteCore()
+                RuntimePackAssets.Select(item => RuntimePackAssetInfo.FromItem(item));
+
+
+            ProjectContext projectContext = lockFile.CreateProjectContext(
+                NuGetUtils.ParseFrameworkName(TargetFramework),
+                RuntimeIdentifier,
+                PlatformLibraryName,
+                RuntimeFrameworks,
+                IsSelfContained);
+
+            var builder = new DependencyContextBuilder2(mainProject, projectContext, IncludeRuntimeFileVersions);
+
+            builder = builder
+                .WithMainProjectInDepsFile(IncludeMainProject)
+                .WithReferenceAssemblies(referenceAssemblyInfos)
+                .WithDirectReferences(directReferences)
+                .WithDependencyReferences(dependencyReferences)
+                .WithReferenceProjectInfos(referenceProjects)
+                .WithExcludeFromPublishAssets(excludeFromPublishAssets)
+                .WithRuntimePackAssets(runtimePackAssets)
+                .WithCompilationOptions(compilationOptions)
+                .WithReferenceAssembliesPath(FrameworkReferenceResolver.GetDefaultReferenceAssembliesPath())
+                .WithPackagesThatWereFiltered(GetFilteredPackages());
+
+            if (CompileReferences.Length > 0)
+            {
+                builder = builder.WithCompileReferences(ReferenceInfo.CreateReferenceInfos(CompileReferences));
+            }
+
+            var resolvedNuGetFiles = ResolvedNuGetFiles.Select(f => new ResolvedFile(f, false))
+                                .Concat(ResolvedRuntimeTargetsFiles.Select(f => new ResolvedFile(f, true)));
+            builder = builder.WithResolvedNuGetFiles(resolvedNuGetFiles);
+
+            DependencyContext dependencyContext = builder.Build();
+
+            var writer = new DependencyContextWriter();
+            using (var fileStream = File.Create(depsFilePath))
             {
                 writer.Write(dependencyContext, fileStream);
             }
-            _filesWritten.Add(new TaskItem(DepsFilePath));
+            _filesWritten.Add(new TaskItem(depsFilePath));
+        }
 
+        protected override void ExecuteCore()
+        {
+            if (DepsFileGenerationMode.Equals("old", StringComparison.InvariantCultureIgnoreCase))
+            {
+                WriteDepsFileOld(DepsFilePath);
+            }
+            else if (DepsFileGenerationMode.Equals("new", StringComparison.InvariantCultureIgnoreCase))
+            {
+                WriteDepsFileNew(DepsFilePath);
+            }
+            else
+            {
+                var newDepsFilePath = Path.ChangeExtension(DepsFilePath, ".new.json");
+
+                WriteDepsFileOld(DepsFilePath);
+
+                WriteDepsFileNew(newDepsFilePath);
+
+                var oldJson = JObject.Parse(File.ReadAllText(DepsFilePath));
+                var newJson = JObject.Parse(File.ReadAllText(newDepsFilePath));
+
+                newJson.Should().BeEquivalentTo(oldJson);
```

Comment:
Agreed, I don't think we should do that either.  I should change it to a simple string comparison, or if that doesn't work something that doesn't bring in a new dependency.

Reply:
Taking into account that you called out the perf of this is not necessarily a main line concern, but I wonder if `WriteDepsFileOld` and `WriteDepsFileNew` should take a stream instead of the path.  In the case of `new` and `old`, we write to a file stream.  In the case of `both`, we write to memory streams and compare them, then only write one of them to disk?

Reply:
Oh and I second @rainersigwald's comment regarding the use of fluent.  I don't know what the proper message should be, however.  Even a diff of two deps files might be too verbose for a message.

Reply:
Didn't review anything, but let's not write turd files to disk if we keep the comparison. Even without perf, it's sloppy. Also, I believe it is writing straight to the final output dir? This will cause junk left behind if we crash, etc.

Reply:
If we don't write it to disk, I think we should emit a diagnostic message containing the old and new content that we can get in a log (probably not in the exception message; who wants to see that in their build output when they run into a bug?).'

My concern is that we're going to fail the build when the deps file isn't what we expect, and users won't be able to provide us with something we can use to diagnose the problem.

Reply:
And, yes, no Fluent assertions in the product.

While you're in here, I'd also recommend getting rid of the new Builder().WithX(x).WithY(x).WithZ(z).Build(). It has always irritated me. It is Fluent for the sake of fashion. It is a glorified function call with a bunch of parameters. But you don't need to fix this now. Just getting it off my chest while we're here. LOL.

Reply:
I'll let you guys decide, but I think new only with an escape hatch is sufficient.

Reply:
I'd also be comfortable "enough" with "new by default with escape hatch" as well, but the "both" model does have the benefit of ensuring compatibility correctness, such that we don't erroneously publish a user's application with a bad deps file, breaking them at runtime.  That scares me.

Reply:
We have a lot of previews to go. If the app breaks, we'll hear about it and tell them to use the escape hatch while we fix it. The real interesting thing I think the comparison buys us is finding differences in .deps.json that  don't lead to obvious consequences. That said, I trust you guys to make the call, having invested the effort in writing and reviewing it.